### PR TITLE
Bump minimum python version to 3.10.

### PR DESCRIPTION
**Pull request by janiversen**
Updated all places I could find mentioning python version 3.8 and 3.9 to 3.10

Removed CI pytest 3.9.

### CODE REVIEW

**Comment by janiversen** [comment]
hmmm...interesting that CI does not use ci.yml from the PR, that seems to be a problem in the CI setup. How can you test CI changes if you need to merge them first.

**Comment by kingosticks** [comment]
You should be able to test them in your fork on GitHub.

**Comment by janiversen** [comment]
@jodal with this you can use '|', we should also change the pyright setup so that Optional is not required..

The code becomes less bloated when you write:
```
    parameter: bool = None
```
and not
```
   parameter: Optional[bool] = None
```

WDYT?

**Comment by janiversen** [comment]
@kingosticks same problem in my fork...CI uses the workflow on develop not the workflow in the pull request. I had this problem in pymodbus a long time ago so I will see what I did and suggest it.

**Comment by kingosticks** [comment]
I thought it used exactly the version in your repo, that's weird... It's not really a problem in this case. Getting CI working is a pain

**Review by kingosticks — APPROVED** [review]
Shame we can't change it all in one place, feels like it should be possible! (Some of the docs stuff not included)

**Comment by kingosticks** [comment]
@jodal do we need to rebuild the packages for Debian 12? Or at least change the install docs claim we support it!

File: docs/changelog.rst [hunk 1]
```diff
@@ -17,7 +17,7 @@ v3.5.0 (UNRELEASED)
 Dependencies
 ------------
 
-- Python >= 3.9 is now required. Python 3.7 and 3.8 are no longer supported.
+- Python >= 3.10 is now required. Python 3.7, 3.8 and 3.9 are no longer supported.
```
Comment (jodal) [inline docs/changelog.rst:20]: I love those Oxford commas!

```suggestion
- Python >= 3.10 is now required. Python 3.7, 3.8, and 3.9 are no longer supported.
```

**Comment by jodal** [comment]
> @jodal with this you can use '|', we should also change the pyright setup so that Optional is not required..

I really like having explicit `None`. I'll see about replacing flake8, isort and friends with ruff one of the next days and use ruff to automatically update our code from `Optional[Foo]` to `Foo | None`.

**Comment by jodal** [comment]
> @jodal do we need to rebuild the packages for Debian 12? Or at least change the install docs claim we support it!
> 
> Ahh, but then that'd bump us beyond Ubuntu LTS. We had this situation before. Can't remember what we put in the docs, if only there was a way....

Uhm, that's a good point. We can't build on Debian 11 any more, so then we build on Debian 12, which has Python 3.11, so we can't assume that package will work on the one year older Ubuntu LTS. Are we happy with requiring Ubuntu 23.04 / Debian 12, or do we want to leave this at Python 3.9 for now?

We've been stable for so long that I think I'm comfortable with bumping all the way to Ubuntu 23.04 / Debian 12 now. We could just call it Mopidy 4.0 to make the huge bump in requirements very obvious. That would also free us up to remove some deprecated APIs...

**Comment by janiversen** [comment]
> > @jodal with this you can use '|', we should also change the pyright setup so that Optional is not required..
> 
> I really like having explicit `None`. I'll see about replacing flake8, isort and friends with ruff one of the next days and use ruff to automatically update our code from `Optional[Foo]` to `Foo | None`.

I use ruff in pymodbus (another project of mine) together with precommit and it works real nice !

**Comment by kingosticks** [comment]
Personally I think I'd prefer to wait for the next LTS rather than cause upgrade hassle for users.

**Comment by kingosticks** [comment]
And I'd prefer a major release have something exciting in it - not sure what that is yet.

**Comment by jodal** [comment]
Closing this, as bumping to 3.10 or later probably won't happen before next year, when Ubuntu 24.04 LTS is released (and Raspberry Pi OS hopefully is based on a Debian with 3.10+ too).

**Comment by janiversen** [comment]
Sounds correct, sorry I forgot it was still open.

**Pull request closed without merging**